File: anchor-reference-server/src/main/java/org/stellar/anchor/reference/client/PlatformApiClient.java
package org.stellar.anchor.reference.client;

import java.io.IOException;
import java.util.HashMap;
import java.util.List;
import okhttp3.HttpUrl;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;
import org.stellar.anchor.api.exception.AnchorException;
import org.stellar.anchor.api.platform.GetTransactionResponse;
import org.stellar.anchor.api.platform.PatchTransactionsRequest;
import org.stellar.anchor.api.platform.PatchTransactionsResponse;
import org.stellar.anchor.auth.AuthHelper;
import org.stellar.anchor.util.AuthHeader;
import org.stellar.anchor.util.OkHttpUtil;

public class PlatformApiClient extends BaseApiClient {
  private final AuthHelper authHelper;
  private final String endpoint;

  public PlatformApiClient(AuthHelper authHelper, String endpoint) {
    this.authHelper = authHelper;
    this.endpoint = endpoint;
  }

  public GetTransactionResponse getTransaction(String id) throws IOException, AnchorException {
    Request request = getRequestBuilder().url(endpoint + "/transactions/" + id).get().build();
    String responseBody = handleResponse(client.newCall(request).execute());
    return gson.fromJson(responseBody, GetTransactionResponse.class);
  }

  public PatchTransactionsResponse patchTransaction(PatchTransactionsRequest txnRequest)
      throws IOException, AnchorException {
    HttpUrl url = HttpUrl.parse(endpoint);
    url =
        new HttpUrl.Builder()
            .scheme(url.scheme())
            .host(url.host())
            .port(url.port())
            .addPathSegment("transactions")
            .build();

    RequestBody requestBody = OkHttpUtil.buildJsonRequestBody(gson.toJson(txnRequest));
    Request request = getRequestBuilder().url(url).patch(requestBody).build();
    Response response = client.newCall(request).execute();
    return gson.fromJson(handleResponse(response), PatchTransactionsResponse.class);
  }

  public HashMap<?, ?> health(List<String> checks) throws IOException, AnchorException {
    HttpUrl url = HttpUrl.parse(endpoint);
    HttpUrl.Builder builder =
        new HttpUrl.Builder()
            .scheme(url.scheme())
            .host(url.host())
            .port(url.port())
            .addPathSegment("health");
    if (checks.size() != 0) {
      String strChecks = String.join(",", checks);
      builder.addQueryParameter("checks", strChecks);
    }

    Request request =
        new Request.Builder()
            .url(endpoint + "/health")
            .header("Content-Type", "application/json")
            .get()
            .build();

    String responseBody = handleResponse(client.newCall(request).execute());
    return gson.fromJson(responseBody, HashMap.class);
  }

  Request.Builder getRequestBuilder() {
    Request.Builder requestBuilder =
        new Request.Builder().header("Content-Type", "application/json");

    AuthHeader<String, String> authHeader = authHelper.createAuthHeader();
    return authHeader == null
        ? requestBuilder
        : requestBuilder.header(authHeader.getName(), authHeader.getValue());
  }
}

File: platform/src/main/kotlin/org/stellar/anchor/platform/PlatformApiClient.kt
package org.stellar.anchor.platform

import okhttp3.HttpUrl.Companion.toHttpUrl
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.stellar.platform.apis.platform.requests.PatchTransactionsRequest
import org.stellar.platform.apis.platform.responses.GetTransactionResponse
import org.stellar.platform.apis.platform.responses.PatchTransactionsResponse

class PlatformApiClient(private val endpoint: String) : SepClient() {
  fun getTransaction(id: String): GetTransactionResponse {
    val request =
      Request.Builder()
        .url("$endpoint/transactions/$id")
        .header("Content-Type", "application/json")
        .get()
        .build()
    val responseBody = handleResponse(client.newCall(request).execute())
    return gson.fromJson(responseBody, GetTransactionResponse::class.java)
  }

  fun patchTransaction(txnRequest: PatchTransactionsRequest): PatchTransactionsResponse? {
    val urlBuilder = this.endpoint.toHttpUrl().newBuilder().addPathSegment("transactions")
    val requestBody = gson.toJson(txnRequest).toRequestBody(TYPE_JSON)
    val request =
      Request.Builder()
        .url(urlBuilder.build())
        .header("Content-Type", "application/json")
        .post(requestBody)
        .build()
    var response = client.newCall(request).execute()
    return gson.fromJson(handleResponse(response), PatchTransactionsResponse::class.java)
  }
}
